gdal/doc/source/development/rfc/rfc76_ogrpythondrivers.rst

```diff
@@ -0,0 +1,472 @@
+   on Windows, look for Python symbols. If found, use it. This is for example
+   the case if GDAL is used from a Python module (GDAL Python bindings, rasterio, etc.)
+   or an application like QGIS that starts a Python interpreter.
+2. otherwise, look for the PYTHONSO environment variable that should point to
+   a pythonX.Y[...].so/.dll
+3. otherwise, look for the python binary in the path and try to identify the
+   correspond Python .so/.dll
+4. otherwise, try to load with dlopen()/LoadLibrary() well-known names of
+   Python .so/.dll
+
+Impacts on GDAL core
+--------------------
+
+They are minimal. The GDALAllRegister() method has an added call to
+GDALDriverManager::AutoLoadPythonDrivers() that implements the above mentioned
+logic. The GDALDriver class has been extended to support a new function
+pointer, IdentifyEx(), which is used by the C++ shim that loads the Python code.
+
+.. code-block:: c++
+
+    int                 (*pfnIdentifyEx)( GDALDriver*, GDALOpenInfo * );
+
+This extended IdentifyEx() function pointer, which adds the GDALDriver* argument,
+is used in priority by GDALIdentify() and GDALOpen() methods. The need for that
+is purely boring. For normal C++ drivers, there is no need to pass the driver,
+as there is a one-to-one correspondence between a driver and the function that
+implements the driver. But for the Python driver, there is a single C++ method
+that does the interface with the Python Identify() method of several Python drivers,
+hence the need of a GDALDriver* argument to forward the call to the appropriate
+driver.
+
+.. _example:
+
+Example of such a driver
+------------------------
+
+Note that the prefixing by the driver name in the connection string is absolutely
+not a requirement, but something specific to this particular driver which is a
+bit artifical. The CityJSON driver mentionned below does not need it.
+
+.. code-block:: python
+
+    #!/usr/bin/env python
+    # -*- coding: utf-8 -*-
+    # This code is in the public domain, so as to serve as a template for
+    # real-world plugins.
+    # or, at the choice of the licensee,
+    # Copyright 2019 Even Rouault
+    # SPDX-License-Identifier: MIT
+
+    # The gdal_python_driver module is defined by the GDAL library at runtime
+    from gdal_python_driver import BaseDriver, BaseDataset, BaseLayer
+
+    # Metadata parsed by GDAL C++ code at driver pre-loading, starting with '# gdal: '
+    # Required and with that exact syntax since it is parsed by non-Python
+    # aware code. So just literal values, no expressions, etc.
+    # gdal: DRIVER_NAME = "DUMMY"
+    # API version(s) supported. Must include 1 currently
+    # gdal: DRIVER_SUPPORTED_API_VERSION = [1]
+    # gdal: DRIVER_DCAP_VECTOR = "YES"
+    # gdal: DRIVER_DMD_LONGNAME = "my super plugin"
+
+    # Optional driver metadata items.
+    # # gdal: DRIVER_DMD_EXTENSIONS = "ext1 est2"
+    # # gdal: DRIVER_DMD_HELPTOPIC = "http://example.com/my_help.html"
```

Comment:
Recommend putting these directives above the Python import statement.

Reply:
done